Fix Overlord leader election when task lock re-acquisition fails

AmatyaAvadhanula opened the pull request:
<!-- Thanks for trying to help us make Apache Druid be the best it can be! Please fill out as much of the following information as is possible (where relevant, and remove it when irrelevant) to help make the intention and scope of this PR clear in order to ease review. -->

<!-- Please read the doc for contribution (https://github.com/apache/druid/blob/master/CONTRIBUTING.md) before making this PR. Also, once you open a PR, please _avoid using force pushes and rebasing_ since these make it difficult for reviewers to see what you've changed in response to their reviews. See [the 'If your pull request shows conflicts with master' section](https://github.com/apache/druid/blob/master/CONTRIBUTING.md#if-your-pull-request-shows-conflicts-with-master) for more details. -->

Fixes Overlord leader election when task lock re-acquisition fails

<!-- Replace XXXX with the id of the issue fixed in this PR. Remove this section if there is no corresponding issue. Don't reference the issue in the title of this pull-request. -->

<!-- If you are a committer, follow the PR action item checklist for committers:
https://github.com/apache/druid/blob/master/dev/committer-instructions.md#pr-and-issue-action-item-checklist-for-committers. -->

### Description

<!-- Describe the goal of this PR, what problem are you fixing. If there is a corresponding issue (referenced above), it's not necessary to repeat the description here, however, you may choose to keep one summary sentence. -->

https://github.com/apache/druid/issues/11653 describes an issue where Overlord leader election fails due lock re-acquisition issues

<!-- Describe your patch: what did you change in code? How did you fix the problem? -->
This PR aims to solve the issue by failing such tasks and clearing all their locks when re-acquisition doesn't succeed, so that the Overlord leader election is not blocked

<!-- If there are several relatively logically separate changes in this PR, create a mini-section for each of them. For example: -->


<!--
In each section, please describe design decisions made, including:
 - Choice of algorithms
 - Behavioral aspects. What configuration values are acceptable? How are corner cases and error conditions handled, such as when there are insufficient resources?
 - Class organization and design (how the logic is split between classes, inheritance, composition, design patterns)
 - Method organization and design (how the logic is split between methods, parameters and return types)
 - Naming (class, method, API, configuration, HTTP endpoint, names of emitted metrics)
-->


<!-- It's good to describe an alternative design (or mention an alternative name) for every design (or naming) decision point and compare the alternatives with the designs that you've implemented (or the names you've chosen) to highlight the advantages of the chosen designs and names. -->

<!-- If there was a discussion of the design of the feature implemented in this PR elsewhere (e. g. a "Proposal" issue, any other issue, or a thread in the development mailing list), link to that discussion from this PR description and explain what have changed in your final design compared to your original proposal or the consensus version in the end of the discussion. If something hasn't changed since the original discussion, you can omit a detailed discussion of those aspects of the design here, perhaps apart from brief mentioning for the sake of readability of this PR description. -->

<!-- Some of the aspects mentioned above may be omitted for simple and small changes. -->

<hr>

##### Key changed/added classes in this PR
 * `TaskLockbox`
 * `TaskQueue`

<hr>

<!-- Check the items by putting "x" in the brackets for the done things. Not all of these items apply to every PR. Remove the items which are not done or not relevant to the PR. None of the items from the checklist below are strictly necessary, but it would be very helpful if you at least self-review the PR. -->

This PR has:
- [x] been self-reviewed.
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java:
```diff
@@ -197,17 +205,26 @@ public int compare(Pair<Task, TaskLock> left, Pair<Task, TaskLock> right)
+        for (TaskLock lock : taskStorage.getLocks(task.getId())) {
+          taskStorage.removeLock(task.getId(), lock);
+        }
+        taskStorage.setStatus(TaskStatus.failure(task.getId(), "Failed to reacquire lock"));
```
This doesn't seem the right thing to do in `TaskLockbox`. If you notice, the `taskMaster` is used in this class to manipulate the locks and that's it. Lockbox doesn't modify any state other than the locks. We could maybe return the list of tasks for which we could not acquire locks to the caller. As you do this, it will be better to return a `Result` object where we can add more info later if needed. what do you think?

AmatyaAvadhanula replied:
done

AmatyaAvadhanula commented:
@abhishekagarwal87 thank you for the review!
I think there were a few other gaps such as cleaning up the task's existing TaskLockPosse and shutting it down if it's running, which I've fixed.

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java:
```diff
@@ -197,7 +206,19 @@ public int compare(Pair<Task, TaskLock> left, Pair<Task, TaskLock> right)
+        }
+      }
+      return new SyncResult(
+          ImmutableList.copyOf(failedToReacquireLockTasks)
```
I think wrapping inside ImmutableList is unnecessary.

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java:
```diff
@@ -197,7 +206,19 @@ public int compare(Pair<Task, TaskLock> left, Pair<Task, TaskLock> right)
+
+      // unlock all TaskLockPosse created, and also remove lock entries from storage
+      for (Task task : failedToReacquireLockTasks) {
+        unlockAll(task);
+        for (TaskLock lock : taskStorage.getLocks(task.getId())) {
+          taskStorage.removeLock(task.getId(), lock);
+        }
```
should this all be in a block where the exceptions can be ignored?

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskMaster.java:
```diff
@@ -416,4 +418,22 @@ public Map<String, Long> getBlacklistedTaskSlotCount()
+  private void processSyncResult(SyncResult syncResult, TaskStorage taskStorage, TaskRunner taskRunner)
+  {
+    String taskReacquisitionFailure = "Failed to reacquire lock";
+    for (Task task : syncResult.getTasksToFail()) {
+      // Shutdown task if it is running
+      taskRunner.shutdown(task.getId(), taskReacquisitionFailure);
+      // Mark as failed
+      taskStorage.setStatus(TaskStatus.failure(task.getId(), taskReacquisitionFailure));
```
should we ignore any exception in this block?

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/SyncResult.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.List;
+
+/**
+ * Result of TaskLockbox#syncFromStorage()
```
nit-pick - This description could be improved.

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskMaster.java:
```diff
@@ -113,7 +113,7 @@ public void becomeLeader()
 
         try {
-          taskLockbox.syncFromStorage();
+          SyncResult syncResult = taskLockbox.syncFromStorage();
```
This is interesting. `TaskLockbox` is synced with storage just once when we become leader, whereas the `TaskQueue` is synced periodically?

I wonder if we shouldn't just move the invocation of `taskLockbox.syncFromStorage()` to `TaskQueue.start()`. It would help `TaskQueue` ensure that `TaskLockbox` is actually synced without having to rely on something done before it was constructed.
This would also help us avoid passing the `SyncResult` around.

The only difference I see is if `taskLockbox.syncFromStorage()` throws an exception, then currently we would fail early, while becoming leader. Whereas, if we move the sync inside `TaskQueue`, it would fail on the lifecycle start, which is effectively the same.

AmatyaAvadhanula replied:
Thanks! I have moved both TaskLockbox's sync and the processing to TaskQueue's start

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/SyncResult.java:
```diff
@@ -0,0 +1,46 @@
+ * Result of TaskLockbox#syncFromStorage()
+ * Contains tasks which need to be forcefully failed to let the overlord become the leader
+ */
+class SyncResult
```
Maybe rename to `TaskLockboxSyncResult`

AmatyaAvadhanula replied:
Done

kfaraz reviewed (changes requested):
Added some comments.

AmatyaAvadhanula commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java:
```diff
@@ -770,6 +776,17 @@ private void syncFromStorage()
           addTaskInternal(task);
         }
 
+        for (Task task : tasksToFail) {
```
Thanks. I've moved this

AmatyaAvadhanula commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java:
```diff
@@ -183,20 +184,33 @@ public int compare(Pair<Task, TaskLock> left, Pair<Task, TaskLock> right)
-              "Could not reacquire lock on interval[%s] version[%s] for task: %s",
+          failedToReacquireLockTaskGroups.add(task.getGroupId());
+          log.error(
+              "Could not reacquire lock on interval[%s] version[%s] for task: %s.",
```
Done

AmatyaAvadhanula commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java:
```diff
@@ -858,6 +873,26 @@ public void unlockAll(Task task)
     }
   }
 
+  public void unlockUnacquiredLocks(Task task)
```
Ah sorry. Removed it

AmatyaAvadhanula commented:
@kfaraz thank you for the review!

AmatyaAvadhanula commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java:
```diff
@@ -108,17 +108,20 @@ public TaskLockbox(
 
     try {
       // Load stuff from taskStorage first. If this fails, we don't want to lose all our locks.
-      final Set<String> storedActiveTasks = new HashSet<>();
```
Thanks for pointing this out. I've reverted the change

kfaraz commented on indexing-service/src/test/java/org/apache/druid/indexing/overlord/http/OverlordTest.java:
```diff
@@ -270,19 +300,20 @@ public void testOverlordRun() throws Exception
     // should return number of tasks which are not in running state
     response = overlordResource.getCompleteTasks(null, req);
-    Assert.assertEquals(2, (((List) response.getEntity()).size()));
+    Assert.assertEquals(4, (((List) response.getEntity()).size()));
```
Why did we need to change an existing test?
I would advise adding a separate test for verifying the behaviour of such tasks where we failed to reacquire locks.

kfaraz commented on indexing-service/src/test/java/org/apache/druid/indexing/overlord/TaskLockboxTest.java:
```diff
@@ -1258,6 +1259,39 @@ public void testGetLockedIntervalsForRevokedLocks() throws Exception
+  @Test
+  public void testFailedToReacquireTaskLock() throws Exception
+  {
+    final Task badTask0 = NoopTask.withGroupId("BadTask");
```
Probably a better name explaining why it's bad or good?

AmatyaAvadhanula replied:
I added a TestLockbox class which returns null for tasks whose group contains "BadTask". I'm not sure it's the best approach to test these changes.
I'm hoping you could take a look and suggest a better approach.

kfaraz replied:
The testing approach itself is fine but the names can be made a little more self-explanatory,
like "TaskWithFailingLockAcquisition" or something. Name the Task instances also accordingly.

AmatyaAvadhanula replied:
Done

kfaraz commented on indexing-service/src/test/java/org/apache/druid/indexing/overlord/TaskLockboxTest.java:
```diff
@@ -1258,6 +1259,39 @@ public void testGetLockedIntervalsForRevokedLocks() throws Exception
+    testLockbox.add(badTask1);
+    testLockbox.add(goodTask0);
+
+    testLockbox.tryLock(badTask0, new TimeChunkLockRequest(TaskLockType.EXCLUSIVE,
```
Nit: Style: Put each arg on a separate line (when necessary) for consistency with the rest of Druid code.

AmatyaAvadhanula replied:
Done

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java:
```diff
@@ -228,6 +235,13 @@ public ScheduledExecutors.Signal call()
           }
       );
       requestManagement();
+      // Remove any unacquired locks from storage
```
Thanks for adding the comment!
I am a little unclear on why there would be unacquired locks left behind. I would imagine that `shutdown()` would take care of this. If not, please rephrase the comments to clarify that point.

AmatyaAvadhanula replied:
shutdown() only clears tasklocks for which a TaskLockPosse is present. The error here is that a lock could not be reacquired and a TaskLockPosse doesn't exist for the conflicting task, which is why these "unacquired" entries must be removed in this manner

kfaraz replied:
Okay, please include this info in the comments.

AmatyaAvadhanula replied:
Included

kfaraz reviewed (changes requested):
Thanks for the quick fix, @AmatyaAvadhanula !
I have added some more minor feedback. I would also request you to test this out thoroughly (in case you haven't already) on a Druid cluster as this changes task queue startup, which affects all task executions.

AmatyaAvadhanula commented:
@kfaraz, the changes have been tested on a druid cluster by introducing a bad entry in the metadata store to simulate this error and I monitored the changes for a while, and have ticked the corresponding box in the checklist.

Thank you for the additional feedback. I'll be sure to test the changes on a cluster after addressing these comments as well.

AmatyaAvadhanula commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java:
```diff
@@ -173,6 +174,12 @@ public void start()
+      // Clean up needs to happen after tasks have been synced from storage
+      Set<Task> tasksToFail = taskLockbox.syncFromStorage().getTasksToFail();
+      for (Task task : tasksToFail) {
+        shutdown(task.getId(), "Failed to reacquire lock.");
```
Thanks for the suggestion. Shouldn't it be "while becoming leader"?

AmatyaAvadhanula replied:
Done

AmatyaAvadhanula commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java:
```diff
@@ -770,6 +776,17 @@ private void syncFromStorage()
 
+        for (Task task : tasksToFail) {
+          try {
+            tasks.putIfAbsent(task.getId(), task);
```
With the previous changes, it was. It has been fixed while addressing your comments

AmatyaAvadhanula commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java:
```diff
@@ -183,20 +187,39 @@ public int compare(Pair<Task, TaskLock> left, Pair<Task, TaskLock> right)
+      }
+
+      Set<Task> tasksToFail = new HashSet<>();
+      for (Task task : taskStorage.getActiveTasks()) {
```
Fixed

kfaraz reviewed (approved):
+1 after CI passes.
